File: sdks/python/bolt_prediction_sdk/core/retry.py
"""
重试模块
"""

import asyncio
from typing import Callable, Awaitable, Any, List
from loguru import logger


class RetryManager:
    """重试管理器"""

    def __init__(
        self,
        max_retries: int = 3,
        backoff_factor: float = 0.5,
        status_codes: List[int] = None,
    ):
        self.max_retries = max_retries
        self.backoff_factor = backoff_factor
        self.status_codes = status_codes or [429, 500, 502, 503, 504]

    async def execute(
        self,
        func: Callable[..., Awaitable[Any]],
        *args,
        **kwargs,
    ) -> Any:
        """
        执行带重试的异步函数

        Args:
            func: 要执行的异步函数
            *args: 位置参数
            **kwargs: 关键字参数

        Returns:
            函数执行结果
        """
        last_exception = None

        for attempt in range(self.max_retries + 1):
            try:
                return await func(*args, **kwargs)
            except Exception as e:
                last_exception = e
                status_code = getattr(e, "status_code", None)

                if status_code not in self.status_codes:
                    raise

                if attempt >= self.max_retries:
                    logger.warning(
                        f"Max retries ({self.max_retries}) reached, giving up"
                    )
                    raise

                wait_time = self.backoff_factor * (2 ** attempt)
                logger.warning(
                    f"Retry attempt {attempt + 1}/{self.max_retries}, "
                    f"waiting {wait_time}s before retry. Error: {e}"
                )
                await asyncio.sleep(wait_time)

        raise last_exception

File: sdks/typescript/src/core/retry.ts
/**
 * 重试模块
 */

export interface RetryManagerOptions {
  maxRetries?: number;
  backoffFactor?: number;
  statusCodes?: number[];
}

export class RetryManager {
  private maxRetries: number;
  private backoffFactor: number;
  private statusCodes: number[];

  constructor(options: RetryManagerOptions = {}) {
    this.maxRetries = options.maxRetries ?? 3;
    this.backoffFactor = options.backoffFactor ?? 0.5;
    this.statusCodes = options.statusCodes ?? [429, 500, 502, 503, 504];
  }

  /**
   * 执行带重试的异步函数
   */
  async execute<T>(fn: () => Promise<T>): Promise<T> {
    let lastError: Error | null = null;

    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      try {
        return await fn();
      } catch (error: any) {
        lastError = error;

        const statusCode = error?.response?.status || error?.statusCode;
        if (!this.statusCodes.includes(statusCode)) {
          throw error;
        }

        if (attempt >= this.maxRetries) {
          console.warn(`Max retries (${this.maxRetries}) reached, giving up`);
          throw error;
        }

        const waitTime = this.backoffFactor * Math.pow(2, attempt) * 1000;
        console.warn(
          `Retry attempt ${attempt + 1}/${this.maxRetries}, ` +
          `waiting ${waitTime}ms before retry. Error: ${error.message}`
        );

        await this.sleep(waitTime);
      }
    }

    throw lastError!;
  }

  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}
